registry/remote/auth/scope.go

```diff
@@ -93,7 +119,62 @@ func AppendScopes(ctx context.Context, scopes ...string) context.Context {
 // GetScopes returns the scopes in the context.
 func GetScopes(ctx context.Context) []string {
 	if scopes, ok := ctx.Value(scopesContextKey{}).([]string); ok {
-		return append([]string(nil), scopes...)
+		return slices.Clone(scopes)
+	}
+	return nil
+}
+
+// scopesPerHostContextKey is the context key for per-host scopes.
+type scopesPerHostContextKey struct{}
+
+// WithScopesPerHost returns a context with per-host scopes added.
+// Scopes are de-duplicated.
+// Scopes are used as hints for the auth client to fetch bearer tokens with
+// larger scopes.
+//
+// For example, uploading blob to the repository "hello-world" does HEAD request
+// first then POST and PUT. The HEAD request will return a challenge for scope
+// `repository:hello-world:pull`, and the auth client will fetch a token for
+// that challenge. Later, the POST request will return a challenge for scope
+// `repository:hello-world:push`, and the auth client will fetch a token for
+// that challenge again. By invoking WithScopesPerHost with the scope
+// `repository:hello-world:pull,push`, the auth client with cache is hinted to
+// fetch a token via a single token fetch request for all the HEAD, POST, PUT
+// requests.
+//
+// Passing an empty list of scopes will virtually remove the scope hints in the
+// context for the given host.
+//
+// Reference: https://docs.docker.com/registry/spec/auth/scope/
+func WithScopesPerHost(ctx context.Context, host string, scopes ...string) context.Context {
```

Comment:
I was struggling with naming... The options were:
1. `WithScopesPerHost`
2. `WithPerHostScopes`
3. `WithScopesForHost`

Reply:
1 sounds good to me.

Reply:
`WithScopesForHost` sounds good to me.

Reply:
I suggest `WithHostScopes`, lol.

Reply:
The votes are tied!